### ISSUE TITLE
How should range literals defined with mixed-type param bounds behave?

### ISSUE THREAD

**Issue by mppf** [chapel-lang/chapel#20545]
Consider code like this:

``` chapel
var r = 1:int(8)..10;
```

What should be the type of `r` ? Before PR #20528, it was `int(8)`, and that was consistent with the behavior of `1:int(8) + 10` which would also result in an `int(8)`.

However, PR #20528 changed the behavior of param-only calls to `+` so that `1:int(8) + 10`  now results in an `int(64)` (just as `myInt8 + myInt64` would produce an `int(64)`).

At the same time, that PR added workarounds in module code so that, specifically for range literals, the old behavior is preserved. As a result, the current behavior after that PR is that `1:int(8)..10;` results in an `int(8)` range.

Should we:
 * keep the behavior for range literals (which now differs from the behavior of `+`), or
 * change the behavior for range literals to be more like the behavior of `+` after PR #20528?

Additionally, should we provide other ways to construct ranges that specify the type more directly? E.g.
 * `new range(idxType=int(8), 1, 10)`
 * `(1..10).castIdxType(int(8))`

**Comment by bradcray** [comment]
Should the title of this issue mention `param` like "range literals defined with mixed-type param bounds"?  Or does the question apply to ranges not bounded by params as well?

**Comment by mppf** [comment]
Only param bounds. I'll update the title, thanks.

**Comment by vasslitvinov** [comment]
I propose to allow casts to generic range and domain types, to mean "preserve the aspects that are generic). This will allow us to write `(1..10).castIdxType(int(8))` as `(1..10): range(int(8), ?)`.

**Comment by bradcray** [comment]
I think when we originally ran into this issue, I was feeling a bit gun-shy about changing the behavior, because we've definitely had users complain about the effort required to get a range of `int(8)`s (say), and I felt like this would be a step backwards in terms of productivity.  However, I think I also may have missed (or too quickly thought past) the fact that this would primarily affect ranges where both bounds were params and not ones like `1..myUint8ConstOrVar` (right?).

I think param-param ranges where a user wants the bounds to be `[u]int(8)` do come up in practice and that casting both bounds (`1:uint(8)..10:uint(8)`) feels a bit onerous, so I don't my current perspective will completely resolve the tension; but I think I agree that consistency trumps convenience here.

I like Vass's suggestion to rely on a range cast (though it has the downside of throwing currently, so would need to be appropriately `try`'d; or else we'd want to go back to making range casts non-throwing and relying on safeCast variations to throw after 2.0 if this became annoying).  Ultimately, I'd prefer to be able to write this as:

```chapel
var r8: range(int(8)) = 1..10;
```

where today that causes problems due to the lack of param ranges.  Adding those would be a big lift, but something we want anyway, so in the foreseeable future.

**Comment by mppf** [comment]
> that this would primarily affect ranges where both bounds were params and not ones like `1..myUint8ConstOrVar` (right?).

Yes, it would only affect ranges where both bounds are params.

I have a slight leaning towards keeping the old behavior, because it provides a relatively convenient way to express the type of a loop induction variable e.g. with `for i in 1:int(8)..10`. AFAIK, we have the long-term intention to support `for i: int(8) in 1..10` which is arguably a better way to write it, but I’m expecting that it will be a while before we get to implementing that.

**Comment by bradcray** [comment]
Hmmmm...  The long-term intention is how I'd prefer to write that pattern.  Do you envision that it'll be awhile before we get to it because nothing has motivated us to do it so far?  I was imagining/hoping that maybe it would be relatively straightforward to implement, particularly in a dyno front-end world...  Put another way, I'd prefer not to support the asymmetry simply because we haven't gotten to this yet (given how attractive and orthogonal of a concept typing the loop index variable is).

**Comment by mppf** [comment]
It's mostly just that we have a lot cooking before the Fall release with the stabilization candidate. I'm also comfortable making the opposite choice and trying to get `for i: int(8) in 1..10` done soon (but still, probably after the Fall release).

**Comment by bradcray** [comment]
I'm not imagining it would need to be done for the fall release.  If users complain about having to either do the double-cast or declare the range separately for such cases, that'll just give us the motivation we finally need to get this feature done (I've had a few other motivating cases over the years, including maybe a shootout (?), but not enough to get me over the hump of trying it.

**Comment by bradcray** [comment]
One of the things I worry about with the status quo is the complexity of describing the behavior of what idxType you get out of applying a `..` operator to two bounds based on their types and param-ness.  Whereas by keeping things consistent, we can either describe it as being like any other built-in integral x integral => integral operator or (someday, ultimately) implement and document the `..` operator as we do others.

**Comment by vasslitvinov** [comment]
In yesterday's group meeting, we favored consistency of `..` with `+`. In order to get a range with idxType=int(8), one could either cast both bounds or cast the entire range, i.e., `(1..10):range(int(8))`. The latter will no longer throw once we implement the decisions in https://github.com/chapel-lang/chapel/issues/19269#issuecomment-1678271195.

While we favored consistency, some of us were surprised by the current behavior of `+`, which was implemented in #20528 and can be summarized in this table:

| a | b | a+b | today's idxType of a..b | "consistent" idxType of a..b |
| --- | --- | --- | --- | --- |
| myInt8 | myInt64 | int(64) | int(64) |  " |
| 1: int(8) | myInt64 | int(64) | int(64) |  " |
| myInt8 | 10 | int(8) | int(8) |  " |
| 1: int(8) | 10 | int(64) | int(8) | int(64) |

We also expressed the desire to make it easy to use smaller integers in loops and ranges for code running on GPUs. We still chose the "consistent" option because only a small amount of extra effort is needed to create ranges with smaller idxTypes.

**Comment by vasslitvinov** [comment]
My w.i.p. implementation of the above change brought up an interesting question: what should happen when a multi-dimensional domain literal has different-idxType ranges?

```chpl
var D = { range1: range(int(8)), range2: range(uint(8)) };
```

Currently this is disallowed. We can allow it post-2.0 as desired.

**Comment by vasslitvinov** [comment]
Resolved and implemented in #23218.
